hazelcast/src/main/java/com/hazelcast/spi/impl/operationparker/impl/WaitSet.java

```diff
@@ -82,6 +85,18 @@ public void park(BlockingOperation op) {
         }
     }
 
+    /**
+     * Returns {@code true} if the operation is parked in this wait set
+     */
+    private boolean isParked(BlockingOperation op) {
+        for (WaitSetEntry entry : queue) {
```

Comment:
Maybe we should add a property on the operation, to avoid iterating over this queue every time?
Won't this lead to some increased congestion?

Reply:
True. This is just one idea for a fix. Ideally we want to avoid the operation being parked twice and also being run twice (when it returns `false` from `shouldWait`). This possibly requires two properties. Or maybe not. I'm not sure at the moment :)